addons/mail/tests/common.py

```diff
@@ -1259,3 +1260,48 @@ def _generate_attachments_data(count, res_model, res_id, attach_values=None, pre
             'res_id': res_id,
             **attach_values,
         } for x in range(count)]
+
+    def _add_time_to_json(self, track_json, record, time_to_add, now):
+        track_json[str(record.stage_id.id)] = track_json.get(str(record.stage_id.id), 0) + \
+            int(time_to_add.total_seconds())
+        now.return_value += time_to_add
+
+    def _assert_change_stage(self, time_to_add, new_stage, track_json, now):
+
+        # We spend time_to_add in the current stage
+        self._add_time_to_json(track_json, self.rec_1, time_to_add, now)
+        self.rec_1.stage_id = new_stage
+        self.flush_tracking()
+
+        # We spend 5 minutes in the new stage
+        self._add_time_to_json(track_json, self.rec_1, timedelta(minutes=5), now)
+        self.rec_1._compute_json_status_time()
+        self.assertEqual(track_json, self.rec_1.json_status_time)
+
+        # And again 5 minutes (ensures that current stage time is computed dynamically)
+        self._add_time_to_json(track_json, self.rec_1, timedelta(minutes=5), now)
+        self.rec_1._compute_json_status_time()
+        self.assertEqual(track_json, self.rec_1.json_status_time)
+
+    def _add_time_to_jsons(self, minutes, now, dic_trackings, batch):
+        time_to_add = timedelta(minutes=minutes)
+        now.return_value += time_to_add
+        for dic_tracking, record in zip(dic_trackings, batch):
+            dic_tracking[str(record.stage_id.id)] = dic_tracking.get(str(record.stage_id.id), 0) + \
+                int(time_to_add.total_seconds())
+
+    def _assert_batch_status_time(self, dic_trackings, batch, new_stage, now):
+
+        # Add 10 minutes spent in the current stage
+        self._add_time_to_jsons(10, now, dic_trackings, batch)
+
+        # Change stage
+        batch.stage_id = new_stage
+        self.flush_tracking()
+
+        # Add 5 minutes spent in the new current stage
+        self._add_time_to_jsons(5, now, dic_trackings, batch)
+        batch._compute_json_status_time()
+        self.assertEqual(dic_trackings[0], batch[0].json_status_time)
+        self.assertEqual(dic_trackings[1], batch[1].json_status_time)
+        self.assertEqual(dic_trackings[2], batch[2].json_status_time)
```

Comment:
Hey just a small comment for the test part :slightly_smiling_face:
I would put this in a different class, `MockStatusbarTime` perhaps ? And import it in the test with `MailCommon`.
I would also rename the methods to match the other, so `assertBatchStatusTime` and `assertChangeState` ?
would be great to have a small docstring in each of them
Also I think that we could merge the `add_time_to_json(s)`, so something like this:
```python
def _add_time_to_json_trackings(self, json_trackings, minutes_to_add, records, now):
    """ Utility method to add time in the json simulating the statusbar time widget.
        :param json_trackings<list>: list of dict where the time info are stored for each stage
        :param minutes_to_add<int>: time in minutes to add in the time tracking of stage
        :param records<recordset>: the records that are impacted
        :param now: patch.object for self.env.cr.now
    """
    time_to_add = timedelta(minutes=minutes_to_add)
    for json_tracking, record in zip(json_trackings, records):
        json_tracking[str(record.stage_id.id)] = json_tracking.get(str(record.stage_id.id), 0) + \
            int(time_to_add.total_seconds())
    now.return_value += time_to_add
```
Cheers!